### ISSUE TITLE
Consider safer access to deeply nested object properties

### ISSUE THREAD

**Issue by ivkos** [Schmavery/facebook-chat-api#553]
Currently, [_formatAttachment](https://github.com/Schmavery/facebook-chat-api/blob/master/utils.js#L231) and [formatMessage](https://github.com/Schmavery/facebook-chat-api/blob/master/utils.js#L517) parse received attachments and messages by directly trying to access deeply nested properties of the objects (e.g.: `blob.story_attachment.media.image.width`). This has the consequence that if any of the properties in this chain has a value of null/undefined, the access will fail with a TypeError.

One way to achieve safer access to such properties is to check if each one of them is non-null and defined, e.g.: 
```js
if (blob && blob.story_attachment && blob.story_attachment.media && blob.story_attachment.media.image) { 
    return blob.story_attachment.media.image.width;
} else { 
    return undefined;
}
```
Another alternative is [_.get(obj, path)](https://lodash.com/docs/4.17.4#get) by lodash which allows us to do this:
```js
_.get(blob, 'story_attachment.media.image.width')
```
and this returns undefined if any of the properties cannot be accessed.

I want to point out that this issue is not a bug report per se, but rather a request for consideration, since implementing this suggestion may make changes made by Facebook harder to spot and debugging harder.

However, if we evaluate which message properties are integral and absolutely necessary to be returned to the users of the API and which are somewhat okay to be missing or to return a default value (until there's a fix), we can make those format functions less fragile.

**Comment by mitchcapper** [comment]
I don't disagree with this, anything that could be null should be checked and JS does not make that easy.   I started with exporting the .get functionality from lodash to see about replicating that, but honestly was over 100 lines and still was not done with all the imports.  

Here is a simplified version I think would work well for our use:
```
/*
	Call SafeGet to get a value that may have a parent that is null.  Similar to lodash get.
	SafeGet supports two calling syntaxes:
		SafeGet(obj,"par1","par2","leaf");
		SafeGet(obj,"par1.par2.leaf");
*/
function SafeGet(obj /*, args */){
	if (obj == null)
		return undefined;
	var path = Array.prototype.slice.call(arguments, 1);
	if (path.length == 0)
		return undefined;
	if (path.length == 1 && path[0].indexOf("."))
		path = path[0].split(".");
	let index = 0;
	const length = path.length;
	while (obj != null && index < length) {
		obj = obj[path[index++]];
	}
	return (index && index == length) ? obj : undefined;
}
```
assuming @Schmavery agrees with the approach I can put a PR together. There is some performance hit,  lodash does cache the figuring out of the path, however ours is simpler and so I am not sure would benefit greatly from caching.

I would also recommend we move to allowing for 'failed' parsing of attachments.  Maybe try/catch the entire thing and if failed create an attachment_failed object with the attachment source and the exception message rather than the infinite loop it ends up in now.

**Comment by ivkos** [comment]
@mitchcapper There is the npm package [lodash.get](https://www.npmjs.com/package/lodash.get) that only has the _.get function exported.

**Comment by mitchcapper** [comment]
@ivkos it is almost 1K lines of code, it handles a lot of use cases but I am not sure it is worth importing something so large for our simple cases.

**Comment by Schmavery** [comment]
Hey guys, I read this issue wrong the first time and didn't realize that the api is failing "in an infinite loop". Can someone clarify what happens? Many exceptions get thrown but the app doesn't crash completely? One way or another that shouldn't be happening, we only want one error per exception.

There are some great suggestions in this thread. I think that first things first, the main issue that's standing out to me is that we're getting "spammed" with errors that we can't catch (as far as I understand).  I think that at a high level, we should be not throwing from calls like `listen()`, instead catching any exceptions and calling the callback with an error. I thought this was already the behaviour but maybe something slipped up. If we do that, then unrecognized exceptions will still cause errors, but they should be ones that you can handle in your code (and then tell us about!).

I can understand the motivation behind wanting default values for fields.  However, in our experience, this leads to the api being incredibly difficult to maintain and debug because facebook can quietly change the format whenever they want.  We rely on our amazing users to help us notice when this happens and update the formatting functions in one place so that they don't have to update their own code.
This attempt to format attachments was a decision we made recently to try and give that same experience for attachments.  We formerly didn't format them at all and made users figure out how they were laid out. The ramp up to full support is a little rocky as we try and figure out what all the possible attachments are but hopefully, with all of your help, it can end up improving the experience overall

**Comment by mitchcapper** [comment]
So I would have to check my code.  It probably isn't the api that infinite loops but I am guessing:
1) Api throws an exception when parsing the attachment of a message
2) My code auto-restarts the connection on error
3) The API re-reads the same message and trys to parse it again
So technically it is my code causing the loop but in reality:
*) We can catch the error.  Yes we shouldn't have any exceptions in parseAttachments if we 100% understand the API.  As you mention however FB can change the API so when they do I think it would be better to set an attachment_parse_failed field and provide the raw attachment object as another field.  Yes this is not useful for code that is expecting attachments, but would allow them to continue to operate until the library gets updated.
*) I agree that if we add something like `SafeGet` to prevent null accessors we may not detect API changes as fast as we throw fewer exceptions.  We would still get  bug reports that say image_width is set to undefined rather than the proper size.  On the plus side a small API change on facebooks part to a field name would not result in the entire attachment being unable to be parsed.  I see the argument both ways.

It is great to have errors/api changes noticed right away but it would also be fantastic if the library did not catastrophically fail in an unrecoverable way on any change.

**Comment by Schmavery** [comment]
I think we're on the same page here. The listen callback takes 2 arguments, an error and a message. In the case where anything in the parsing of a message fails, we should catch the exception and call the callback with the error (which can contain whatever we want, including the original attachment).
People should already be checking this error value, so an issue in parsing an attachment shouldn't cause the app to "catastrophically fail in an unrecoverable way on any change".

**Comment by mitchcapper** [comment]
Ok so in checking my code I only restart the connection if the error status code is 400 (normally re-login) so the looping behavior is library driven.  If I understand your proposal it is to try/catch around the message parsing.  On failure call the callback with an error object containing the original exception and  the message that couldn't be parsed.   

Correct?

**Comment by Schmavery** [comment]
That's right

**Comment by ivkos** [comment]
@mitchcapper I've just tried your workaround (i.e. wrapping the formatDeltaMessage call in try/catch and calling the globalCallback with the error) and it works. 

Although I don't really understand how listen() works, I think the underlying issue might be related to how listen() gets called recursively. But it would take way too much effort to refactor and will be hard to test.

**Comment by Schmavery** [comment]
@ivkos thanks a lot for trying it out. Can you make a quick PR with the change? I'm happy to explain more about how listen() works and drill deeper into this but would be nice to at least fix the bug for now.

**Comment by ivkos** [comment]
@Schmavery #554

**Comment by Schmavery** [comment]
Closed for now by #554 and https://github.com/Schmavery/facebook-chat-api/commit/7552ce69561ac424cf6461e4c0cd8762ae2a7b9f. Thanks all.
